store-implementation/simple-federated-store/src/main/java/uk/gov/gchq/gaffer/federated/simple/FederatedUtils.java

```diff
@@ -0,0 +1,153 @@
+                && ((OperationView) operation).getView() != null
+                && ((OperationView) operation).getView().hasGroups()) {
+
+            // Update the view for the graph
+            ((OperationView) operation).setView(
+                getValidViewForGraph(((OperationView) operation).getView(), graphSerialisable));
+
+            updatedOperations.add(operation);
+
+        // Recursively go into operation chains to make sure everything is fixed
+        } else if (operation instanceof OperationChain) {
+            for (final Operation op : ((OperationChain<?>) operation).getOperations()) {
+                // Resolve if haven't hit the depth limit for validation
+                if (depth < 5) {
+                    updatedOperations.addAll(getValidOperationForGraph(op, graphSerialisable, depth + 1).getOperations());
+                } else {
+                    LOGGER.warn("Hit depth limit of 5 whilst making the operation valid for graph. The View may be invalid for Graph: {}", graphSerialisable.getGraphId());
+                    updatedOperations.add(op);
+                }
+            }
+        } else {
+            updatedOperations.add(operation);
+        }
+
+        // Create and return the fixed chain for the graph
+        OperationChain<?> newChain = new OperationChain<>();
+        newChain.setOptions(operation.getOptions());
+        newChain.updateOperations(updatedOperations);
+
+        return newChain;
+    }
+
+    /**
+     * Returns a {@link View} that contains groups only relevant to the graph. If
+     * the supplied view does not require modification it will just be returned.
+     *
+     * @param view The view to make valid.
+     * @param graphSerialisable The relevant graph.
+     * @return A version of the view valid for the graph.
+     */
+    public static View getValidViewForGraph(final View view, final GraphSerialisable graphSerialisable) {
+        final Schema schema = graphSerialisable.getSchema();
+
+        // Figure out all the groups relevant to the graph
+        final Set<String> validEntities = new HashSet<>(view.getEntityGroups());
+        final Set<String> validEdges = new HashSet<>(view.getEdgeGroups());
+        validEntities.retainAll(schema.getEntityGroups());
+        validEdges.retainAll(schema.getEdgeGroups());
+
+        if (!validEntities.equals(view.getEntityGroups()) || !validEdges.equals(view.getEdgeGroups())) {
+            // Need to make changes to the view so start by cloning the view
+            // and clearing all the edges and entities
+            final View.Builder builder = new View.Builder()
+                    .merge(view)
+                    .entities(Collections.emptyMap())
+                    .edges(Collections.emptyMap());
+            validEntities.forEach(e -> builder.entity(e, view.getEntity(e)));
+            validEdges.forEach(e -> builder.edge(e, view.getEdge(e)));
+            return builder.build();
```

Comment:
What happens here if the corrected view is empty? Can we end up in a situation where we forward an empty view onto a graph which may retrieve unexpected results? E.g. if I did GetAll 'created' edges from a federated graph of the Modern dataset, would the 'knows' graph receive an empty view and then return its knows edges?

Reply:
That is a good point I don't think that situation was handled on old federation either I guess in that situation you would probably want to skip execution on that graph all together. Maybe could throw an exception to state that nothing specified in the View is relevant to that graph then the user can choose whether to resubmit the chain but with a skip on fail option or more targeted graph selection.